### DISCUSSION TITLE
Module not found: Can't resolve imported dependency "fs"

### DISCUSSION THREAD

**Discussion by test4appdev1** [quasarframework/quasar#11869] [General - Components / Directives / etc] [answered]
I got this error below:

```
Module not found: Can't resolve imported dependency "fs"
Did you forget to install it? You can run: yarn add fs
```

Then I installed "fs" but there is still the error.

`yarn add fs`

Then, to solve the error, I installed "node-polyfill-webpack-plugin":

`yarn add -D node-polyfill-webpack-plugin`

Then, I configured the plugin in quasar.conf.js:

```
const NodePolyfillPlugin = require('node-polyfill-webpack-plugin');
module.exports = configure(function (ctx) {
  return {
    build: {
      extendWebpack(cfg) {
        cfg.plugins.push(new NodePolyfillPlugin({}));
      }
    }
  }
}
```

But there is still the error. Are there any solutions to solve the error?

**Comment by MilosPaunovic** [comment]
Have you followed the steps from here:
[https://quasar.dev/start/upgrade-guide#nodejs-polyfills](https://quasar.dev/start/upgrade-guide#nodejs-polyfills)

**Reply by test4appdev1** [reply]
Yes, I did as I've already shown it on my first post. I have one idea. Do you think downgrading webpack 5 to webpack 4 solves this problem?

**Reply by MilosPaunovic** [reply]
As you can see, your code and the one from the docs aren't identical.

Have you tried exactly as in the docs:
```js
// quasar.conf.js
build: {
  chainWebpack (chain) {
    const nodePolyfillWebpackPlugin = require('node-polyfill-webpack-plugin')
    chain.plugin('node-polyfill').use(nodePolyfillWebpackPlugin)
  }
}
```

**Reply by test4appdev1** [reply]
Sorry, I misunderstood but I tried your way as well below https://quasar.dev/start/upgrade-guide#nodejs-polyfills then could not solve the problem:

`yarn add --dev node-polyfill-webpack-plugin
`
```
// quasar.conf.js
build: {
  chainWebpack (chain) {
    const nodePolyfillWebpackPlugin = require('node-polyfill-webpack-plugin')
    chain.plugin('node-polyfill').use(nodePolyfillWebpackPlugin)
  }
}
```

Actually, I'm trying to use "Shopify Library" then I got the error. 
Do you think downgrading webpack 5 to webpack 4 solves this problem?

**Reply by hawkeye64** [reply]
> Do you think downgrading webpack 5 to webpack 4 solves this problem?

Absolutely not. If you want to use webpack 4.x, then you will have to use Quasar v1.
I believe, after the above changes are made, there is still a setting when creating the BrowserWindow that is needed.

**Reply by test4appdev1** [reply]
OK. I've just downgraded webpack 5 to webpack 4 then couldn't solve the problem as **hawkeye64** said. Are there any ways to solve this problem?

**Reply by pdanpdan** [reply]
To be honest I think you are trying to use the shopify-node-api client side :)

**Reply by test4appdev1** [reply]
This is the client shopify library which I use: 

`yarn add @shopify/shopify-api
`

**Reply by pdanpdan** [reply]
Yes, as expected - that is https://github.com/Shopify/shopify-node-api

**Reply by test4appdev1** [reply]
No, it's not.

**Reply by pdanpdan** [reply]
https://www.npmjs.com/package/@shopify/shopify-api
See the GitHub url of the repo
See here: https://github.com/Shopify/shopify-node-api/issues/147

**Comment by lucacalcaterra** [comment]
same issue here...

**Reply by MilosPaunovic** [reply]
Have you read through the responses on a tread above?

**Comment by CynthiaBlue** [comment]
I'm having the same problem. I've installed npm i node-polyfill-webpack-plugin and added the code to the quasar.conf..js file and I'm still getting a bunch of errors when I try to run Nodemailer.

**Comment by yzqdev** [comment]
## using Electron or NodeJS API in the rederer process is strongly discouraged, better not do like this
The same issue. i use quasar to build an electron app. The `ipcRenderer` function need to require fs module, so it comes out with the error `Module not found: Can't resolve imported dependency "fs"`

like this
```vue
<script setup lang="ts">
import { ipcRenderer } from "electron";
function miniBtnClick() {
  ipcRenderer.on("asynchronous-reply", (event, arg) => {
    console.log(arg); // prints "pong"
  });
}
</script>
```
i've installed  `node-polyfill-webpack-plugin` and added the code to the `quasar.conf.js` file according to [https://quasar.dev/start/upgrade-guide#nodejs-polyfills](https://quasar.dev/start/upgrade-guide#nodejs-polyfills), error still exists.

**Comment by MilosPaunovic** [comment, marked as answer]
Take a look at this issue [https://github.com/electron/electron/issues/9920](https://github.com/electron/electron/issues/9920).

**Reply by yzqdev** [reply]
thx, that's my fault

**Comment by CynthiaBlue** [comment]
I was able to get nodemailer to work by putting it into the electron preload file. This is what I did. [https://github.com/quasarframework/quasar/discussions/12545#discussioncomment-2232021](https://github.com/quasarframework/quasar/discussions/12545#discussioncomment-2232021)

**Comment by cashewpillar** [comment]
I used Quasar CLI and when i incorporate dotenv or env.json I get the warning
```
Module not found: Can't resolve imported dependency "fs"
Did you forget to install it? You can run: yarn add fs
```

 despite exactly following https://quasar.dev/start/upgrade-guide#nodejs-polyfills
```javascript
chainWebpack (chain) {
        const nodePolyfillWebpackPlugin = require('node-polyfill-webpack-plugin')
        chain.plugin('node-polyfill').use(nodePolyfillWebpackPlugin)
        
        chain.plugin('eslint-webpack-plugin')
          .use(ESLintPlugin, [{ extensions: [ 'js', 'vue' ] }])

      }
```

**Reply by cashewpillar** [reply]
https://github.com/ChainSafe/web3.js/issues/4169 

Is there an alternative to setting up environment variables? (just for auth0 client secrets)

**Reply by metalsadman** [reply]
https://quasar.dev/quasar-cli-vite/handling-process-env#using-dotenv

https://www.npmjs.com/package/dotenv
